File: src/main/java/fi/quanfoxes/lexer/OperatorToken.java
package fi.quanfoxes.lexer;

import java.util.Objects;

public class OperatorToken extends Token {

    private Operator operator;

    public OperatorToken(String identifier) {
        super(TokenType.OPERATOR);
        this.operator = Operators.get(identifier);
    }

    public OperatorToken(Operator operator) {
        super(TokenType.OPERATOR);
        this.operator = operator;
    }

    public Operator getOperator ()  {
        return operator;
    }

    @Override
    public String getText() {
        return operator.getIdentifier();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof OperatorToken)) return false;
        if (!super.equals(o)) return false;
        OperatorToken that = (OperatorToken) o;
        return operator == that.operator;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), operator);
    }
}

File: Zigzag/Lexer/OperatorToken.cs
using System;
using System.Collections.Generic;

public class OperatorToken : Token
{
	public Operator Operator { get; private set; }

	public OperatorToken(string identifier) : base(TokenType.OPERATOR)
	{
		Operator = Operators.Get(identifier);
	}

	public OperatorToken(Operator operation) : base(TokenType.OPERATOR)
	{
		Operator = operation;
	}

	public override bool Equals(object? obj)
	{
		return obj is OperatorToken token &&
			   base.Equals(obj) &&
			   EqualityComparer<Operator>.Default.Equals(Operator, token.Operator);
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(base.GetHashCode(), Operator);
	}

	public override object Clone()
	{
		return MemberwiseClone();
	}
}
